packages/core/src/repository/cache.repository.ts

```diff
@@ -54,7 +56,28 @@ export class CacheRepository<T> implements GetRepository<T>, PutRepository<T>, D
             case MainOperation:
                 return this.getMain.get(query);
             case CacheOperation:
-                return this.getCache.get(query);
+                return this.getCache.get(query).then((value: T) => {
+                    if (value == null) {
+                        throw new NotFoundError();
+                    }
+                    if (!this.validator.isObjectValid(value)) {
+                        throw new NotValidError();
+                    }
+                    return value;
+                }).catch((err: Error) => {
+                    if (err instanceof NotValidError || err instanceof NotFoundError) {
+                        return this.get(query, new MainOperation()).catch((finalError: Error) => {
+                            let op = operation as CacheOperation;
+                            if (op.fallback(finalError)) {
+                                return this.getCache.get(query);
```

Comment:
What I don't understand is, if `getCache` in line 59 has already failed, why this is going to work now? `MainOperation` doesn't update the cache.

Reply:
I must be missing something, as `CacheSyncOperation` does the same thing. In that case I see the same… `getCache` fails, we try to `MainSyncOperation` (get main, update cache)… and in case of error we try to `getCache`. But, if we're here is because `getMain` has failed, and so does `getCache`…

I don't understand anything. xD

Reply:
I've changed the code. I think it will make more sense now.

Reply:
Ok, that makes more sense. We should apply the same fix for `CacheSyncOperation`, no?

Reply:
Maybe @joselufo or @vilanovi can give us some insight here. 

For the `CacheOperation` I'm checking within the `CacheRepository` if `err instanceof NotValidError`, and if true we check the fallback lambda. Maybe this `err instanceof NotValidError` check should be the responsibility of the fallback lambda. I'm actually doing it this way because I know that if the error is `NotFoundError` there is no point in re-trying, we can throw the error directly 

Regarding the `CacheSyncOperation`, the first fallback is the Main data source. If this still fails we fallback to the lambda, not doing any further check. This implementation was already in place, I didn't change anything. Also checking the Kotlin implementation and it is very similar : https://github.com/mobilejazz/harmony-kotlin/pull/33/files